lib/srv/desktop/rdp/rdpclient/src/rdpdr/mod.rs

```diff
@@ -350,75 +833,690 @@ impl Client {
+            // into an RDP data structure. Because of how next() is implemented for FileCacheObject,
+            // the first time this is called we will get an object for the "." directory, the second
+            // time will give us "..", and then we will iterate through any files/directories stored
+            // within dir.
+            if let Some(fso) = dir.next() {
+                match req.file_info_class_lvl {
+                // TODO(isaiah): we should support all the file_info_class_lvl's that FreeRDP does:
+                // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L794
+                FileInformationClassLevel::FileBothDirectoryInformation => {
+                    let buffer = Some(FileInformationClass::FileBothDirectoryInformation(
+                        FileBothDirectoryInformation::from(fso)?
+                    ));
+                    self.prep_drive_query_dir_response(
+                        &req.device_io_request,
+                        NTSTATUS::STATUS_SUCCESS,
+                        buffer
+                    )
+                },
+                FileInformationClassLevel::FileFullDirectoryInformation => {
+                    let buffer = Some(FileInformationClass::FileFullDirectoryInformation(
+                        FileFullDirectoryInformation::from(fso)?
+                    ));
+                    self.prep_drive_query_dir_response(
+                        &req.device_io_request,
+                        NTSTATUS::STATUS_SUCCESS,
+                        buffer
+                    )
+                }
+                FileInformationClassLevel::FileDirectoryInformation |
+                FileInformationClassLevel::FileNamesInformation => {
+                    Err(not_implemented_error(&format!(
+                        "support for ServerDriveQueryDirectoryRequest with file_info_class_lvl = {:?} is not implemented",
+                        req.file_info_class_lvl
+                    )))
+                },
+                _ => {
+                    Err(invalid_data_error("received invalid FileInformationClassLevel in ServerDriveQueryDirectoryRequest"))
+                }
+            }
+            } else {
+                // Once our iterator is exhausted, send back a NTSTATUS::STATUS_NO_MORE_FILES to alert RDP that we've listed all the
+                // contents of this directory.
+                // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/winpr/libwinpr/file/generic.c#L1193
+                // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_main.c#L114
+                self.prep_drive_query_dir_response(
+                    &req.device_io_request,
+                    NTSTATUS::STATUS_NO_MORE_FILES,
+                    None,
+                )
+            }
+        } else {
```

Comment:
This is a very "top heavy" if.

I would invert it, just like we commonly do in Go. Check for the error case first, return an error, and then "unindent" the rest of the logic.

Reply:
2e41c37aa5bd30003c916de98f4f677809471d1c